### ISSUE TITLE
Unable to import layer from Postgis view or materialized view , if it has no no primary key

### ISSUE THREAD

**Issue by wmeinolf** [qgis/QGIS#52943]
### What is the bug or the crash?

I have various Materialized Views in Postgis, to reduce number of attributes of source table and to speed up access. 

When loaded into layer table with drad&drop from browser, I get attention sign, on mouse over the message, that the layer was not found (invalid Layer not loaded). On rightclick the only option is to repair layer, which does not fix the layer. 
Next I tried to open Layer in DB Manager, which is possible without problems. On import from DB Manager same like with drag&drop.
Next I tried to add layer via menu. Here I found losts of attention signs, with message, to define primary key and at right side the set tickbox, that primary key is available. Import was unsuccessful, same like drag& drop. Once I unticked the primary key box, I was able to successfully import the layer to the layer tab.

All mat views were defined quite a while ago and worked perfectly unter 3.16. Unter 3.28.x .Windows and 3.22.x Ubuntu the issue occurs same way. 

Older QGIS Projects generated under 3.16 had a entirely different source definition, which may explain the bug. They obviously generated the missing primary key via SQL:

pre 3.22
service='bbvcrmmirror' key='_uid0_' checkPrimaryKeyUnicity='0' table="(SELECT row_number() over () AS _uid0_, * FROM (SELECT * FROM \"public\".\"nok_baulose_mat_view\" ) AS _subq_0_ )" (geometry)

current 3.28
Valid: 
service='bbvcrmmirror' sslmode=disable key='Kennung - Baulos' srid=25832 type=Point checkPrimaryKeyUnicity='0' table="public"."nok_baulose_mat_view" (geometry)

invalid (drag&drop)
service='bbvcrmmirror' sslmode=disable key='"Kennung - Baulos"' srid=25832 type=Point checkPrimaryKeyUnicity='1' table="public"."nok_baulose_mat_view" (geometry)

 

### Steps to reproduce the issue

Load Materialized View, which does not have primary key from PostgreSQL in browser by double click or drag&drop

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS-Version | 3.28.5-Firenze | QGIS-Codeversion | 50adba36f2
-- | -- | -- | --
Qt-Version | 5.15.3
Python-Version | 3.9.5
GDAL-Version | 3.6.3
PROJ-Version | 9.2.0
EPSG-Registraturdatenbankversion | v10.082 (2023-02-06)
GEOS-Version | 3.11.2-CAPI-1.17.2
SQLite-Version | 3.41.1
PDAL-Version | 2.5.2
PostgreSQL-Client-Version | unknown
SpatiaLite-Version | 5.0.1
QWT-Version | 6.1.6
QScintilla2-Version | 2.13.1
BS-Version | Windows 10 Version 2009
  |   |   |  
Aktive Python-Erweiterungen
disconnected-islands | 2.0.2
GeoCoding | 2.19
inspireatomclient | 0.8.1
lizmap | 3.12.0
mmqgis | 2021.9.10
ORStools | 1.5.3
Qgis2threejs | 2.7.1
QuickOSM | 2.2.2
quick_map_services | 0.19.33
wfsclient | 0.9.11
db_manager | 0.1.20
grassprovider | 2.12.99
MetaSearch | 0.3.6
processing | 2.12.99
sagaprovider | 2.12.99

</body></html>QGIS-Version
3.28.5-Firenze
QGIS-Codeversion
[50adba36f2](https://github.com/qgis/QGIS/commit/50adba36f2)
Qt-Version
5.15.3
Python-Version
3.9.5
GDAL-Version
3.6.3
PROJ-Version
9.2.0
EPSG-Registraturdatenbankversion
v10.082 (2023-02-06)
GEOS-Version
3.11.2-CAPI-1.17.2
SQLite-Version
3.41.1
PDAL-Version
2.5.2
PostgreSQL-Client-Version
unknown
SpatiaLite-Version
5.0.1
QWT-Version
6.1.6
QScintilla2-Version
2.13.1
BS-Version
Windows 10 Version 2009

Aktive Python-Erweiterungen
disconnected-islands
2.0.2
GeoCoding
2.19
inspireatomclient
0.8.1
lizmap
3.12.0
mmqgis
2021.9.10
ORStools
1.5.3
Qgis2threejs
2.7.1
QuickOSM
2.2.2
quick_map_services
0.19.33
wfsclient
0.9.11
db_manager
0.1.20
grassprovider
2.12.99
MetaSearch
0.3.6
processing
2.12.99
sagaprovider
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

**Comment by wmeinolf** [comment]
![Issue](https://user-images.githubusercontent.com/3882351/235309281-fa3b0cd6-3569-4176-bb71-b6da5c50622a.jpg)
![Issue_unticked](https://user-images.githubusercontent.com/3882351/235309287-0e725534-fc54-4291-9eee-edac74d11f90.jpg)

**Comment by wmeinolf** [comment]
Hi, yes "Kennung - Baulos" is the intended UID. In the meantime I use the following workaround: 
I first create a view, which contains all relevant columns collected from various tables records. "Kennung - Baulos"  is always unique!  When I then create or refresh the materialized view, it now containes the additional column "id" with content "row_number() OVER () AS id" generated as UID instead of not working "Kennung - Baulos". Now ID is marked as UID and I can import the layer without problems with drag & drop.

My point is, that peviously it always worked with drag & drop. Due to the workaround in now worka again from QGIS user point of view.

**Comment by troopa81** [comment]
The only difference I see in uri  with QGIS 3.28 is that checkPrimaryKeyUnicity is activated when it fails.

That would mean that "Kennung - Baulos" is actually not unique. To check it we execute the following request

```sql
SELECT count(distinct ("Kennung - Baulos"))=count(("Kennung - Baulos")) FROM "public"."nok_baulose_mat_view"
```

Does that return true on your side? 

If yes, could you share a sample project and data so I can reproduce. If not, well, the key is not unique.

**Comment by wmeinolf** [comment]
Hi and thanks for your help. Indeed the request returned fail. I found multiple appearences so that the key was not unique. I read again my "pre 3.22 Statement at Top:

_pre 3.22
service='bbvcrmmirror' key='uid0' checkPrimaryKeyUnicity='0' table="(SELECT row_number() over () AS uid0, * FROM (SELECT * FROM "public"."nok_baulose_mat_view" ) AS subq_0 )" (geometry)_

I did not realize before, that it is pretty sililar to my current "workaround". Perhaps pre 3.22 it was automatically checked, if the key is uid and, if not, the UID was autogenerated. Anyway, Thanks again for your help. Feel free, to close the case.
